### PR TITLE
[interceptors/validator] feat: add error logging in validator

### PR DESCRIPTION
**Pull request by rohanraj7316**
used logging.Logger interface to
add error logging in validator
interceptor

addition: #494

### CODE REVIEW

File: interceptors/validator/validator.go [hunk 1]
```diff
@@ -83,9 +90,9 @@ func UnaryServerInterceptor(all bool) grpc.UnaryServerInterceptor {
 // returns ALL validation error as a wrapped multi-error.
 // Note that generated codes prior to protoc-gen-validate v0.6.0 do not provide an all-validation
 // interface. In this case the interceptor fallbacks to legacy validation and `all` is ignored.
-func UnaryClientInterceptor(all bool) grpc.UnaryClientInterceptor {
+func UnaryClientInterceptor(all bool, logger logging.Logger) grpc.UnaryClientInterceptor {
```
Comment (bwplotka) [inline interceptors/validator/validator.go:93]: I would definitely create optional argument here and everywhere else we use logger.

So users can do:
`UnaryClientInterceptor()`
`UnaryClientInterceptor(validator.WithLogger(...))`
`UnaryClientInterceptor(validator.WithFailFast())`

We can use this opportunity to clear out what all means (it's simply FailFast).

This is last moment before full v2.0 so we can (and should) break compatibility.

Would you be able to help with that? 🙈

Comment (rohanraj7316) [reply interceptors/validator/validator.go:93]: do we need to make similar changes for other interceptions too?

Comment (bwplotka) [reply interceptors/validator/validator.go:93]: Only validator interceptors (all interceptors in validator package)

Comment (bwplotka) [reply interceptors/validator/validator.go:93]: Otherwise the PR is perfect (:

**Comment by rohanraj7316** [comment]
have updated the implementation do let me your feedback. if it's cool so, will update the doc and test cases.

File: interceptors/validator/options.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+// Logger
+type Logger func() (logging.Level, logging.Logger)
+
+// DefaultLoggerMethod
+func DefaultLoggerMethod() (logging.Level, logging.Logger) {
+	return "", nil
+}
+
+// WithLogger
+func WithLogger(logger Logger) Option {
```
Comment (bwplotka) [inline interceptors/validator/options.go:47]: Why not here:

```suggestion
func WithLogger(logging.Level logger Logger) Option {
```

Simpler to use (:

File: interceptors/validator/options.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// Decision
+type Decision bool
+
+// Decider function defines rules for suppressing any interceptor logs.
+type Decider func() Decision
+
+// DefaultDeciderMethod
+func DefaultDeciderMethod() Decision {
```
Comment (bwplotka) [inline interceptors/validator/options.go:60]: We don't need this, we have `selector` interceptor.

File: interceptors/validator/options.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	return false
+}
+
+// WithFailFast
+func WithFailFast(d Decider) Option {
```
Comment (bwplotka) [inline interceptors/validator/options.go:65]:
```suggestion
// WithFailFast  tells validator to immediately stop doing further validation after first validation error.
func WithFailFast() Option {
```

File: interceptors/validator/validator.go [hunk 5]
```diff
@@ -50,78 +63,60 @@ func validate(req any, all bool) error {
...
+	o := evaluateClientOpt(opts)
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		if err := validate(req, o.shouldFailFast, o.logger); err != nil {
 			return err
 		}
 		return invoker(ctx, method, req, reply, cc, opts...)
 	}
 }
 
-// StreamServerInterceptor returns a new streaming server interceptor that validates incoming messages.
```
Comment (bwplotka) [inline interceptors/validator/validator.go:95]: You can leave majority of this comment e.g especially protoc-gen-validate info still applies.

**Review by bwplotka — COMMENTED** [review]
Amazing work! Generally great direction, just some nits.

File: interceptors/validator/options.go (hunk 2 above)
Comment (rohanraj7316) [reply interceptors/validator/options.go:47]: oops, yaa i might have overthink it!!

Comment (rohanraj7316) [reply interceptors/validator/options.go]: made the changes made on the suggestions.

File: interceptors/validator/validator.go (hunk 5 above)
Comment (rohanraj7316) [reply interceptors/validator/validator.go:95]: updated the comments. also, slight modification in the older ones

**Comment by rohanraj7316** [comment]
@bwplotka can you help me to figure out what's the lint issue?

**Comment by bwplotka** [comment]
Just run "make lint" locally and commit changes (some automatic fixes got applied)

File: interceptors/validator/options.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+func evaluateClientOpt(opts []Option) *options {
+	optCopy := &options{}
+	*optCopy = *defaultOptions
+	for _, o := range opts {
+		o(optCopy)
+	}
+	return optCopy
+}
+
+// WithLogger tells validator to log all the error.
```
Comment (bwplotka) [inline interceptors/validator/options.go:39]:
```suggestion
// WithLogger tells validator to log all the validation errors with the given log level.
```

File: interceptors/validator/options.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+func WithLogger(level logging.Level, logger logging.Logger) Option {
+	return func(o *options) {
+		o.logger = logger
```
Comment (bwplotka) [inline interceptors/validator/options.go:42]: Level not recorded (:

Comment (rohanraj7316) [reply interceptors/validator/options.go:46]: 🤦🏼

**Comment by bwplotka** [comment]
make lint again please until it does not show any more errors locally (:

**Review by bwplotka — APPROVED** [review]
LGTM! Just lint. Thanks!

I can also just merge and fix lint on my own, that's fine - I plan to do one more thing.

**Comment by rohanraj7316** [comment]
sure go ahead actually, I tried running lint at my end but it's missing the initialization of `GOIMPORTS` in env

**Comment by bwplotka** [comment]
What is an error? Because GOIMPORTS is defined in Makefile variable file in .bingo/Variables.mk

Anyway, merging, thanks!

**Pull request merged**